Add custom input fields for editing metadata

aswallace opened the pull request:
This is a follow-up to #338 that adds custom fields for different data types based on [the Figma designs](https://www.figma.com/design/oNQvOAK11YINcbPvNmkgwO/Public-File-Explorer?node-id=3138-10352&node-type=frame&t=FxEJXP0Ufqmnsazm-0).

### Description
The input type switch is modeled off the way we did it in AnnotationFilterForm and re-uses some of the same components.
- Uses our existing choice group/text field/date field components for boolean/string/date+datetime respectively
- Adds a new "duration" form type
- Creates a new numerical input component to be more stylistically consistent (and replaced usage in our NumberRangePicker component)

### Screenshots
Duration 
<img width="300" alt="new duration form" src="https://github.com/user-attachments/assets/718fd23c-3e11-4a21-978c-0967ffc6b442">
Boolean
<img width="300" alt="choice group" src="https://github.com/user-attachments/assets/a6f23685-5f0b-4eac-988d-e8643891deab">
Date/Date Time
<img width="300" alt="date picking form" src="https://github.com/user-attachments/assets/33c06857-c939-461e-96f2-362c5b53628d">
Dropdown
<img width="300" alt="combo box" src="https://github.com/user-attachments/assets/8114ea87-387b-4b92-a808-92d9616759e0">
Number
<img width="300" alt="new numerical input" src="https://github.com/user-attachments/assets/e4b4698d-08c1-42c0-bc4c-61cadb40ecea">
Text/Default
<img width="300" alt="default for any type" src="https://github.com/user-attachments/assets/651205b2-beb3-4208-8c85-c5c8b713f450">

lynwilhelm commented:
@aswallace Great work! Just 3 feedback notes:

1. Duration: The labels are overlapping the inputs. There should be a couple pixels of space in between.
2. Hint text should be consistent with pattern, which is to describe what the user should do using an action word. For example "Type an option name..." So, Text/default should read "Type in value(s)" and Number should say "Type a numerical value..."
3. Can the little calendar be the same color as text so it doesn't look disabled?

Thanks for sharing!

aswallace commented:
Thank you @lynwilhelm! Addressed feedback in [6c52646](https://github.com/AllenInstitute/biofile-finder/pull/354/commits/6c52646965a65d329630a4d0ffcdb9340c04a135) and updated the screenshots to reflect

kmitcham commented on packages/core/components/DurationForm/index.tsx:
```diff
@@ -0,0 +1,73 @@
+                    defaultValue={days}
+                />
+                <NumberField
+                    aria-label="Hours"
```
Is there code somewhere to check that hours < 24, minutes < 60, and seconds < 60?

aswallace replied:
Not currently, but I can add min/max restrictions on the UI

aswallace replied:
added in https://github.com/AllenInstitute/biofile-finder/pull/354/commits/1da0c7f16e2d21319ba164760615e2b8a5c7a1f7

kmitcham commented on packages/core/components/EditMetadata/MetadataDetails.tsx:
```diff
@@ -41,20 +52,95 @@ export default function EditMetadataDetailsList(props: DetailsListProps) {
+        if (column?.fieldName === "value") {
+            if (!fieldContent) return "[No value] (blank)";
+            else return annotationFormatter.displayValue(fieldContent);
+        } else if (column?.fieldName === "fileCount") {
```
so fileCount will always have a value?

aswallace replied:
it should. I'll add a check for clarity though

pgarrison commented on packages/core/components/EditMetadata/MetadataDetails.module.css:
```diff
@@ -62,3 +67,64 @@
+}
+
+.no-padding {
+  padding-bottom: 0 !important;
```
Leave a comment about why the `!important` is necessary?

aswallace replied:
these importants were not, in fact, important. Remnants of UI debugging. Removed in [b1d134a](https://github.com/AllenInstitute/biofile-finder/pull/354/commits/b1d134ad9a522f43a9135affb81e83607db8eace)

pgarrison commented on packages/core/entity/AnnotationFormatter/duration-formatter.ts:
```diff
@@ -19,12 +20,22 @@ export default {
 
     valueOf(value: any) {
+        const RANGE_OPERATOR_REGEX = /([0-9]+)D ([0-9]+)H ([0-9]+)M ([0-9]*\.?[0-9]+)S/g;
+        const exec = RANGE_OPERATOR_REGEX.exec(value);
```
Maybe something like `matches` is  more clear than `exec`?

aswallace replied:
fixed in [761fa46](https://github.com/AllenInstitute/biofile-finder/pull/354/commits/761fa46a0c786142ff324d5007c58dc293e24b0b)

pgarrison commented on packages/core/components/NumberRangePicker/NumberField.tsx:
```diff
@@ -0,0 +1,39 @@
+                onChange={props.onChange}
+                placeholder={props?.placeholder}
+                min={props?.min}
+                max={props?.max}
```
From playing with [this example](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number) input, it seems that `min`and `max` control what the increment/decrement buttons do, but not values entered via keyboard. More dangerous than values out of bounds are non-numeric values. What happens if the user enters `10D 2H 100M 10S` into the hour field?

aswallace replied:
I may be misunderstanding the question, but users are not able to enter non-numeric values in this type of input (with the possible exception of `e`). I've added additional in-bounds validation in [b1d134a](https://github.com/AllenInstitute/biofile-finder/pull/354/commits/b1d134ad9a522f43a9135affb81e83607db8eace)

pgarrison replied:
Great! I didn't open BFF, just tested against a generic `<input type="number" />` element

pgarrison commented on packages/core/components/DurationForm/index.tsx:
```diff
@@ -0,0 +1,80 @@
+                    onChange={(event) => setDurationHours(event?.target?.value || "")}
+                    defaultValue={hours}
+                    min={0}
+                    max={24}
```
I think these `max` values are inclusive? In which case these bounds (here and below) should be 23 and 59

aswallace replied:
good call, I'm used to thinking of max as exclusive. Fixed in [b1d134a](https://github.com/AllenInstitute/biofile-finder/pull/354/commits/b1d134ad9a522f43a9135affb81e83607db8eace)

pgarrison reviewed (approved):
This UI looks so nice!

SeanLeRoy commented on packages/core/components/ComboBox/index.tsx:
```diff
@@ -67,7 +66,7 @@ export default function BaseComboBox(props: Props) {
                 rootChecked: styles.comboBoxItemChecked,
             }}
-            useComboBoxAsMenuWidth={props?.useComboBoxAsMenuWidth}
+            useComboBoxAsMenuWidth={true}
```
```suggestion
            useComboBoxAsMenuWidth
```

aswallace replied:
This was deliberate; FluentUI was not registering the property correctly. I'm not able to replicate the bug now though, so will fix & keep an eye on it

SeanLeRoy commented on packages/core/components/DurationForm/index.tsx:
```diff
@@ -0,0 +1,80 @@
+                    className={styles.inputField}
+                    id="durationDays"
+                    label="Days"
+                    onChange={(event) => setDurationDays(event?.target?.value || "")}
```
Seems like this component could exclusively pass around `Number` type values (for cases with leading zeroes it would just not yet return a value AKA `undefined`); NBD to keep this though

aswallace replied:
I agree in theory, but running into some implementation bugs

SeanLeRoy commented on packages/core/components/EditMetadata/EditMetadata.module.css:
```diff
@@ -81,7 +81,7 @@
 .text-field {
   padding-bottom: var(--margin);
-  max-width: 300px;;
+  width: 300px;;
```
```suggestion
  width: 300px;
```

SeanLeRoy commented on packages/core/components/EditMetadata/MetadataDetails.tsx:
```diff
@@ -41,20 +52,95 @@ export default function EditMetadataDetailsList(props: DetailsListProps) {
+    const inputField = () => {
+        switch (props.fieldType) {
+            case AnnotationType.DATE:
+            case AnnotationType.DATETIME:
```
Is there no time selector for `DATETIME` types or is there always a time selector for `DATE` types? Either of those would be bad no?

aswallace replied:
Unfortunately yes, the date picker does not provide a time selector. This issue also exists in the AnnotationFilterForm

aswallace replied:
I'm opening a separate ticket for this because the FluentUI TimePicker is imo too buggy to use (and the API does not match the component), so making our own is just as good of an option but will hold up this ticket

aswallace replied:
added in [9328a44](https://github.com/AllenInstitute/biofile-finder/pull/354/commits/9328a44ecfedb6b82b1fedfbbaf73679dbb37633)
<img width="500" alt="image" src="https://github.com/user-attachments/assets/85556131-ee0e-43ad-a061-3fdda2641d9f" />
